File: SearchArrayOfUnknownSize.cpp
// Time Complexity : O(logN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    int search(const ArrayReader& reader, int target) {
        int right = 2;
        int left = 0;
        while(reader.get(right) < target)
        {
            left = right;
            right = right * 2;
        }
        
        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            if (reader.get(mid) == target) return mid;
            else if (reader.get(mid) > target) right = mid - 1;
            else left = mid + 1;
        }
        return -1;
    }
};

File: search_array_of_unknown_size.py

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        if reader.get(0) == target:
            return 0
        
        left, right = 0, 1
        while reader.get(right) < target:
            left = right
            right <<= 1
        
        while left <= right:
            pivot = left + ((right - left) >> 1)
            num = reader.get(pivot)
            
            if num == target:
                return pivot
            if num > target:
                right = pivot - 1
            else:
                left = pivot + 1
        
        return -1